### ISSUE TITLE
Ship ESM

### ISSUE THREAD

**Issue by abhishiv** [postcss/postcss#1858]
Right now in order to use post css in the browser

- You have to rely on a bundler and stub fs, path modules
- Use a CDN like jspm which takes care of cjs to esm conversion

It would be great if postcss was to ship ESM directly which could be loaded directly in the browser with importmaps. It seems like most of the code would work without the need of fs module - which could in any case be stubbed the way jspm does.

Ref #830 , #913

**Comment by ai** [comment]
We will do it in a next major release in a few years. JS ecosystem is not ready for ESM right now (I can be sure in it as a author of a few popular ESM-only projects like Nano ID).

**Comment by abhishiv** [comment]
Hey @ai 

Thanks for you reply!

The reason for opening this issue is that I'm building an ESM bundler that uses package-lock.json and converts lockfile to importmap.

So while, I am somewhat vaguely aware of difficulties in supporting ESM, can you please expand on it for my gratification?

**Comment by ai** [comment]
> So while, I am somewhat vaguely aware of difficulties in supporting ESM, can you please expand on it for my gratification?

Only ESM app can use ESM library/dependency.

As the result, we need to wait until all bundlers (and all other scripts which run PostCSS) will be converted to ESM.

**Comment by abhishiv** [comment]
Ah, but that doesn't stop postcss from shipping both cjs and esm?

**Comment by ai** [comment]
> Ah, but that doesn't stop postcss from shipping both cjs and esm

It will increase maintability cost a lot. I tried CJS/ESM dual package on Nano ID and it increase complexity a lot and didn’t work on every case (PostCSS users will have a lot of edge cases).

**Comment by abhishiv** [comment]
Ok, now I see what you mean. 

https://github.com/search?q=repo%3Aai%2Fnanoid+esm+dual&type=issues

**Comment by vikaskaliramna07** [comment]
Well you can't import in browser using import postcss from `npm:postcss`, and hope that never happens
it's still better to use cdn services(it they work fine for postcss) that provide compatibility for browser.

**Comment by dufrenb** [comment]
Hi 👋 Any update on this ? since last year there is a much larger compatibility for ESM support!

**Comment by ai** [comment]
I am waiting for native ESM support in Node.js for CJS apps (be able to require ESM lib).

Without it the migration will be disaster (we will break usage for million user to add native browser support needed for a few users).

Dual ESM/CJS system doesn't work on our scale (as Nano ID author I have a lot of experience).

If you want to use PostCSS in browser just use bundler.

Good news, native ESM support for CJS app is coming to Node.js, but I have no idea about when.

**Comment by ai** [comment]
@danielbayley yes, it is the right way, but not the final stop for PostCSS size.

We need to wait until `require(esm)` support will be everywhere (so, end of Node.js 18 support and a decent time from 20.x/22.x2 releases). I will think about it at least in September.

**Comment by janis-me** [comment]
Heyho! Node 18 hit End-of-life and 20/22 are in Maintenance LTS now. Are there any other blockers for this? I  would love to help make a move or add some sponsoring to make this happen. Maybe even just starting with an "experimental" ESM build just to get wheels rolling?
